### PR TITLE
Lesson 17

### PR DESCRIPTION
**Pull request by namelexa**
### Fixed Issues (if relevant)

1. magento/magento2#<issue_number>: Issue title
2. ...

### Manual testing scenarios

1. ...
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: testapi.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
+<?php
+class MagentoAPI
+{
+    protected $_magentoBaseUrl = 'https://kalexey.local/';
```
Comment (chabanenk0) [inline testapi.php]: Didn't work for me with https. While changing to http it began working. Seems some additional configurations are needed with curl to perform https requests. No need to fix, just keep it in mind

File: testapi.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        $response = curl_exec($ch);
+        $response = json_decode($response);
+        return $response;
+    }
+}
+$magento = new MagentoAPI('kalexey', 'kalexey2');
+$sampleRequest = $magento->getSampleRequest(1);
+var_dump($sampleRequest);
+unset($sampleRequest->created_at);
+$sampleRequest->product_name .= ' (CHANGED)';
```
Comment (chabanenk0) [inline testapi.php]: There was no field $sampleRequest->product_name in your response. Please check the fields names, specify another field to check if data is saving

File: app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+{
+    /**
+     * @param Data\AskQuestionInterface $askQuestion
+     * @return mixed
+     */
+    public function save(AskQuestionInterface $askQuestion);
+
+    /**
+     * @param $askQuestionId
+     * @return mixed
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php]:
```
 @return mixed 
```
- this is incorrect annotation. Magento API uses these annotations to perform requests data formating, so please check carefully the annotations for all input and output variables of API classes. Didn't worked for me.

File: app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function getById($askQuestionId);
+
+    /**
+     * @param SearchCriteriaInterface $searchCriteria
+     * @return mixed
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php]: return type of getList method should be not mixed but:
```
* @return \Magento\Framework\Api\SearchResults
```
Type hinting also need to be checked, but API used annotations as information about what data is returned

File: app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function getList(SearchCriteriaInterface $searchCriteria);
+
+    /**
+     * @param Data\AskQuestionInterface $askQuestion
+     * @return mixed
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php]: * @return bool

File: app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function delete(AskQuestionInterface $askQuestion);
+
+    /**
+     * @param $askQuestionId
+     * @return mixed
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php]: @return bool

File: app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php (hunk 3 above)
Comment (chabanenk0) [reply app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php]: the correct return type for is getById (please check the implementation as well)
```
* @return \AKisilenko\ModuleLesson6\Api\Data\AskQuestionInterface
```

Comment (chabanenk0) [reply app/code/AKisilenko/ModuleLesson6/Api/AskQuestionRepositoryInterface.php]: Also input type hinting need to be specified correctly 
```
    /**
     * @param int $askQuestionId
     * @return \AKisilenko\ModuleLesson6\Api\Data\AskQuestionInterface
     */
    public function getById(int $askQuestionId): AskQuestionInterface;

```

File: app/code/AKisilenko/ModuleLesson6/Cron/Status.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Status constructor.
+     * @param LoggerInterface $logger
+     * @param ScopeConfigInterface $scopeConfig
+     * @param CollectionFactory $askQuestionsFactory
+     */
+    public function __construct(
+        LoggerInterface $logger,
+        ScopeConfigInterface $scopeConfig,
+//        Config $config,
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Cron/Status.php]: remove unneeded commented code

File: app/code/AKisilenko/ModuleLesson6/Model/AskQuestion.php [hunk 8]
```diff
@@ -10,33 +10,134 @@
...
  */
-class AskQuestion extends AbstractModel
+class AskQuestion extends AbstractModel implements AskQuestionInterface
 {
     const STATUS_PENDING = 'pending';
     const STATUS_PROCESSED = 'processed';
 
+    public function getId()
+    {
+        return $this->getData('request_id');
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Model/AskQuestion.php]: for some reason in your database the field name was ask_id, but not request_id:
return $this->getData('ask_id');

File: app/code/AKisilenko/ModuleLesson6/Model/AskQuestion.php [hunk 9]
```diff
@@ -10,33 +10,134 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setId($id)
+    {
+        return $this->setData('request_id', $id);
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Model/AskQuestion.php]: return $this->setData('ask_id', $id);

File: app/code/AKisilenko/ModuleLesson6/Model/AskQuestionRepository.php [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+    /**
+     * @param SearchCriteriaInterface $criteria
+     * @return mixed
+     */
+    public function getList(SearchCriteriaInterface $criteria)
+    {
+        $searchResults = $this->searchResultsFactory->create();
+        $searchResults->setSearchCriteria($criteria);
+        $collection = $this->AskQuestionCollectionFactory->create();
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Model/AskQuestionRepository.php]: Typo here: 
 $collection = $this->askQuestionCollectionFactory->create();
(field name begins with lowercase a: askQ.. not AskQ

File: app/code/AKisilenko/ModuleLesson6/Model/AskQuestionRepository.php [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        $this->resource->load($askQuestion, $askQuestionId);
+        if (!$askQuestion->getId()) {
+            throw new NoSuchEntityException(__('Question with id "%1" does not exist.', $askQuestionId));
+        }
+        return $askQuestion;
+    }
+
+    /**
+     * @param SearchCriteriaInterface $criteria
+     * @return mixed
```
Comment (chabanenk0) [inline app/code/AKisilenko/ModuleLesson6/Model/AskQuestionRepository.php]: check and fix annotations and type hinting

File: testapi.php (hunk 1 above)
Comment (namelexa) [reply testapi.php]: thanks

**Pull request merged**